### PR TITLE
Add kafkajs with producer and consumer

### PR DESCRIPTION
**Pull request by Frawless**
This PR adds kafkajs dependency and also simple implementation of blocking producer and consumer.

As part of this PR I also added two tests - one for producing/consuming messages from topic and one fro browse messages. Filtering will be done in the next PR.

This PR also contains several improvements/fixes for navigation to make the navigation reusable from multiple places and also more stable.

TODO - do not merge yet, still need to do final test

### CODE REVIEW

File: lib/config.ts
Comment (mmusil) [inline lib/config.ts:27]: Good idea!

File: tests/kas.spec.ts
Comment (mmusil) [inline tests/kas.spec.ts:93]: Makes sense, it seems that the instance is deleted quicker than the filter is applied.

Comment (Frawless) [reply tests/kas.spec.ts:93]: Correct, I already mentioned that within the previous PR. I will probably revert this before merging and open another PR which will remove it as it is too flaky.

File: tests/kas_messaging.spec.ts
Comment (mmusil) [inline tests/kas_messaging.spec.ts:21]: Don't forget to change these strings before you get the PR to "ready for review" state, please. Nice easter egg though 😄

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [inline tests/kas_messaging.spec.ts:55]: I see that this can work, but I would feel more comfortable if each step of the logic from here to the end is performed "always" (e.g. outside of the `else` starting at line 37) using appropriate checks to identify the existence of each and every element we are creating (there are already some in place).

This will make it easier to handle while debugging, wdyt?

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [inline tests/kas_messaging.spec.ts:76]:
```ts
expect(producerResponse).toBe(true)
```
?

File: tests/kas_with_instance.spec.ts
Comment (andreaTP) [inline tests/kas_with_instance.spec.ts:36]: this should be fixed now.

File: lib/clients.ts
Comment (andreaTP) [inline lib/clients.ts:100]: Shouldn't we check the exact number of messages? `===`

We can also return the number of messages instead of a boolean so that the check can be performed in the actual test body (easier to debug in case of issues with the Playwright console).

File: lib/clients.ts
Comment (Frawless) [reply lib/clients.ts:18]: Both configurations are available to the users so I think we should cover both.

I think it depends if we want to I am to merge this PR before Christmas or not. I will be off from next Tuesday. However, I don't think the change will be too hard so I will check it today, and maybe we can do a final decision later today or tomorrow.?

File: lib/clients.ts
Comment (Frawless) [reply lib/clients.ts:79]: I am not sure if retries make sense. In case the consumer is attached it will eventually consume messages if the messages will be available. In case the consumer won't be able to read from the topic, it will fail anyway. We are not even able to say "okay, consume just 200 messages" with some simple implementation as we do in Java for example. 

Or did you mean retries for consuming the same messages again?

File: lib/config.ts
Comment (Frawless) [reply lib/config.ts:24]: Agree

File: lib/kafka.ts
Comment (Frawless) [reply lib/kafka.ts:6]: To navigate to KafkaInstances through side menu from everywhere. I think we don't have use case in main for it now. Do you think it would be better to have separate functions for it instead of one complex? This will help with maintainability for sure.

File: lib/kafka.ts
Comment (Frawless) [reply lib/kafka.ts:93]: Agree, I just put there notes about it during development. Will open issues today.

File: lib/popup.ts
Comment (Frawless) [reply lib/popup.ts:3]: Makes sense

File: lib/popup.ts
Comment (Frawless) [reply lib/popup.ts:4]: Good idea, I will try it and rewrite

File: lib/sa.ts
Comment (Frawless) [reply lib/sa.ts:6]: I believe so. The link is visible only on the root page. When you are at `Kafka Instances` page for example, `Application and Data Service` is just a text I think.

File: tests/kas.spec.ts
Comment (Frawless) [reply tests/kas.spec.ts:92]: Yes, I put it there just for cleaner debug of my changes. Will revert it for sure.

File: tests/kas_messaging.spec.ts
Comment (Frawless) [reply tests/kas_messaging.spec.ts:16]: I think it should be possible. I will try it and let you know.

File: lib/clients.ts
Comment (Frawless) [reply lib/clients.ts:100]: Ye, will change it.

File: lib/clients.ts
Comment (andreaTP) [reply lib/clients.ts:79]: fair answer 👍 
let's evolve the code when needed.

File: lib/kafka.ts
Comment (andreaTP) [reply lib/kafka.ts:6]: > Do you think it would be better to have separate functions for it instead of one complex? This will help with maintainability for sure.

Yup, agreed 👍

File: lib/kafka.ts
Comment (andreaTP) [reply lib/kafka.ts:38]: 👍

File: lib/kafka.ts
Comment (andreaTP) [reply lib/kafka.ts:93]: 👍

File: lib/popup.ts
Comment (andreaTP) [reply lib/popup.ts:3]: 👍

File: tests/kas.spec.ts
Comment (andreaTP) [reply tests/kas.spec.ts:92]: thanks 👍

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [reply tests/kas_messaging.spec.ts:16]: thanks 👍

File: lib/clients.ts
Comment (andreaTP) [reply lib/clients.ts:100]: 👍

File: lib/kafka.ts
Comment (Frawless) [reply lib/kafka.ts:93]: 1st TODO - https://issues.redhat.com/browse/MGDSTRM-10281
2nd TODO - https://issues.redhat.com/browse/MGDSTRM-10282

File: lib/kafka.ts
Comment (Frawless) [reply lib/kafka.ts:101]: https://issues.redhat.com/browse/MGDSTRM-10283

File: tests/kas.spec.ts
Comment (Frawless) [reply tests/kas.spec.ts:92]: Also opened https://issues.redhat.com/browse/MGDSTRM-10284

File: tests/kas_messaging.spec.ts
Comment (Frawless) [reply tests/kas_messaging.spec.ts:54]: Yes, that's a very good catch!

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [inline tests/kas_messaging.spec.ts:58]: [nitpick] feel free to ignore, this encoding works, I would have written it as something like:

```
try {
  await expect(page.locator('a', { hasText: testTopicName })).toHaveCount(0);
} catch (e) {
  await createKafkaTopic(page, testTopicName);
}
```

Comment (andreaTP) [inline tests/kas_messaging.spec.ts:58]: [nitpick] feel free to ignore, this encoding works, I would have written it as something like:

```ts
try {
  await expect(page.locator('a', { hasText: testTopicName })).toHaveCount(0);
} catch (e) {
  await createKafkaTopic(page, testTopicName);
}
```

File: lib/clients.ts
Comment (andreaTP) [reply lib/clients.ts:106]: Ok, I finally looked into this, and I believe that some variation of this (WARN I have not tested it!):

```ts
  public async consumeMessages(topic: string, expectedMsgCount: number, fromBeginning = true): Promise<number> {
    let msgCount = 0;

    return new Promise((resolve, reject) => {
      try {
        this.kafkaConsumer.connect()
          .then(() => this.kafkaConsumer.subscribe({ topic: topic, fromBeginning: fromBeginning }))
          .then(() => this.kafkaConsumer.run({
              eachMessage: async (messagePayload: EachMessagePayload) => {
                const { topic, partition, message } = messagePayload;
                const prefix = `${topic}[${partition} | ${message.offset}] / ${message.timestamp}`;
                console.log(`- ${prefix} ${message.key}#${message.value} -> ${msgCount}`);
                msgCount++;
                if (msgCount >= expectedMsgCount) {
                  console.log('Received: ' + msgCount);
                  resolve(msgCount);
                }
              }
            }));
      } catch (error) {
        console.log('Error: ', error);
        reject(error);
      }
    });
  }
```

should be the way to go, all in all:

- perform the consuming logic within the lambda of the `new Promise` which is going to be the one that gets returned
- complete the external Promise on the specific condition

you might need to play a bit with the sync/async part of it to get it to work as expected (Kafka.Js is not typed and I haven't tested the code ... ).
Feel free to ping me if you have any blockers implementing it and we can have a second look together.

Comment (Frawless) [reply lib/clients.ts:106]: Looks interesting, I will play with it and get back to you with the results, thanks!

File: lib/clients.ts
Comment (Frawless) [reply lib/clients.ts:18]: I have been a little bit busier than I expected. I played a little bit with it and decided to implement it in the next PR to avoid additional conflicts and delays with this. Opened https://issues.redhat.com/browse/MGDSTRM-10308 .

File: lib/clients.ts
Comment (Frawless) [reply lib/clients.ts:106]: Working fine form my POV so I replaced the code in client configuration.

File: lib/kafka.ts
Comment (Frawless) [reply lib/kafka.ts:102]: I don't think this is an issue. When you type, the whispering will offer you a topic name. In case you click on it or anywhere else, the button is working fine. Only when you type in the field is it disabled till the next action. Well maybe it's an issue. So should I open the issue wdyt?

File: tests/kas_messaging.spec.ts
Comment (Frawless) [reply tests/kas_messaging.spec.ts:58]: Maybe we can do some kind of refactoring of this code occurrence in a separate PR? There are a lot of occurrences similar to this.

File: lib/clients.ts
Comment (andreaTP) [reply lib/clients.ts]: ok 👍

File: lib/kafka.ts
Comment (andreaTP) [inline lib/kafka.ts:21]: can you remove commented code please?

File: lib/popup.ts
Comment (andreaTP) [inline lib/popup.ts:6]: should we add a shorter timeout? this is otherwise going to take 30 seconds per element ...

File: lib/sa.ts
Comment (andreaTP) [reply lib/sa.ts:6]: not a big deal, fine for me 👍

File: lib/topic.ts
Comment (andreaTP) [inline lib/topic.ts:39]: typo? `navige` -> `navigate` ?

File: lib/topic.ts
Comment (andreaTP) [inline lib/topic.ts:56]: `await` ?

File: lib/topic.ts
Comment (andreaTP) [inline lib/topic.ts:53]: a shorter custom timeout maybe?

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [reply tests/kas_messaging.spec.ts:60]: deal 👍

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [inline tests/kas_messaging.spec.ts:95]: should this be in a `finally` block?

File: tests/kas_messaging.spec.ts
Comment (andreaTP) [inline tests/kas_messaging.spec.ts:89]: I still don't get why, if we are producing 100 messages we might be receiving more than that ... is it a wrong configuration of the acknowledgment?
This should be `toBeEqual` I think ...

File: lib/clients.ts
Comment (andreaTP) [reply lib/clients.ts]: 👍

**Review by andreaTP — CHANGES_REQUESTED** [review]
Last details, but it looks in pretty good shape now, excellent work @Frawless !

**Review by andreaTP — APPROVED** [review]
LGTM! 🎉 
Thanks @Frawless !

**Pull request merged**
